src/rgw/driver/sfs/sqlite/sqlite_versioned_objects.cc

```diff
@@ -13,110 +13,429 @@
+      auto max_commit_time_ids = storage.select(
+          columns(&DBVersionedObject::id, max(&DBVersionedObject::commit_time)),
+          where(
+              is_equal(&DBVersionedObject::object_id, object_id) and
+              is_not_equal(
+                  &DBVersionedObject::object_state, ObjectState::DELETED
+              )
+          ),
+          group_by(&DBVersionedObject::id),
+          order_by(&DBVersionedObject::id).desc()
+      );
+      auto found_value = max_commit_time_ids.size() &&
+                         std::get<1>(max_commit_time_ids[0]) != nullptr;
+      if (found_value) {
+        // if not value is found could be, for example, if lifecycle deleted all
+        // non current versions before.
+        auto last_version_id = std::get<0>(max_commit_time_ids[0]);
+        auto last_version =
+            storage.get_pointer<DBVersionedObject>(last_version_id);
+        if (last_version) {
+          ret_value = *last_version;
+        }
+      }
+      transaction.commit();
+    }
+    return ret_value;
+  } catch (const std::system_error& e) {
+    // throw exception (will be caught later in the sfs logic)
+    // TODO revisit this when error handling is defined
+    throw(e);
+  }
+}
+
+uint SQLiteVersionedObjects::add_delete_marker_transact(
+    const uuid_d& object_id, const std::string& delete_marker_id, bool& added
+) const {
+  uint ret_id{0};
+  added = false;
+  try {
+    auto storage = conn->get_storage();
+    auto transaction = storage.transaction_guard();
+    auto max_commit_time_ids = storage.select(
+        columns(&DBVersionedObject::id, max(&DBVersionedObject::commit_time)),
+        where(
+            is_equal(&DBVersionedObject::object_id, object_id) and
+            is_not_equal(&DBVersionedObject::object_state, ObjectState::DELETED)
+        ),
+        group_by(&DBVersionedObject::id),
+        order_by(&DBVersionedObject::id).desc()
+    );
+    // if found, value we are looking for is in the first position of the results
+    // because we ordered descending in the query
+    auto found_value = max_commit_time_ids.size() &&
+                       std::get<1>(max_commit_time_ids[0]) != nullptr;
+    if (found_value) {
+      auto last_version_id = std::get<0>(max_commit_time_ids[0]);
+      auto last_version =
+          storage.get_pointer<DBVersionedObject>(last_version_id);
+      if (last_version &&
+          last_version->object_state == ObjectState::COMMITTED &&
+          last_version->version_type == VersionType::REGULAR) {
+        auto now = ceph::real_clock::now();
+        last_version->version_type = VersionType::DELETE_MARKER;
+        last_version->delete_time = now;
+        last_version->mtime = now;
+        last_version->version_id = delete_marker_id;
+        ret_id = storage.insert(*last_version);
+        added = true;
+        /// only commit if the delete maker was indeed inserted.
+        /// the rest of calls in this transaction are read operations
```

Comment:
I would drop the docstring from these, and make them regular comments.

Reply:
yup, fixed